packages/itwin/tree-widget/src/tree-widget-react/components/trees/categories-tree/internal/CategoriesTreeVisibilityHandler.ts

```diff
@@ -337,335 +294,317 @@ class CategoriesTreeVisibilityHandlerImpl implements HierarchyVisibilityHandler
     );
   }
 
-  private getModelVisibilityStatus({ modelIds }: { modelIds: Id64Array | Id64Set }): Observable<VisibilityStatus> {
-    const result = defer(() => {
-      const viewport = this._props.viewport;
-
-      return from(modelIds).pipe(
-        distinct(),
-        mergeMap((modelId) => {
-          if (!viewport.view.viewsModel(modelId)) {
-            return from(this._idsCache.getModelCategoryIds(modelId)).pipe(
-              mergeMap((categoryIds) => from(this._idsCache.getCategoriesModeledElements(modelId, categoryIds))),
-              getSubModeledElementsVisibilityStatus({
-                parentNodeVisibilityStatus: createVisibilityStatus("hidden"),
-                getModelVisibilityStatus: (modelProps) => this.getModelVisibilityStatus(modelProps),
-              }),
-            );
-          }
-
-          return from(this._idsCache.getModelCategoryIds(modelId)).pipe(
-            mergeMap((categoryIds) => this.getCategoryDisplayStatus({ modelId, categoryIds, ignoreSubCategories: true })),
+  private getModelsVisibilityStatus({ modelIds }: { modelIds: Id64Arg }): Observable<VisibilityStatus> {
+    return from(Id64.iterable(modelIds)).pipe(
+      mergeMap((modelId) => {
+        // For hidden models we only need to check subModels
+        if (!this._props.viewport.view.viewsModel(modelId)) {
+          return this.getSubModels({ modelIds: modelId }).pipe(
+            mergeMap(({ subModels }) => {
+              if (subModels && Id64.sizeOf(subModels) > 0) {
+                return this.getModelsVisibilityStatus({ modelIds: subModels }).pipe(
+                  map((subModelsVisibilityStatus) =>
+                    subModelsVisibilityStatus.state !== "hidden" ? createVisibilityStatus("partial") : createVisibilityStatus("hidden"),
+                  ),
+                );
+              }
+              return of(createVisibilityStatus("hidden"));
+            }),
             mergeVisibilityStatuses,
```

Comment:
do we need to use `mergeVisibilityStatuses` operator inside the `mergeMap`, when we already have it right after the `mergeMap`?